### DISCUSSION TITLE
Consider switching to the node installer for `just`

### DISCUSSION THREAD

**Discussion by zackkrida** [WordPress/openverse#3875] [Ideas]
The `just` task runner used in Openverse, while a very handy universal utility, may not be a common tool installed by all who which to contribute to Openverse. We could consider streamlining this for users by using the [npm version of Just](https://github.com/brombal/just-install). This would bundle just with our project's Node dependencies.

This would, however, introduce one quirk. `Just` would be installed in `node_modules/bin` (maybe not _exactly_, but for illustrative purposes bear with me) and evoked in package.json scripts. To use it in the project root one would have to type `npx just`. 

Perhaps, then, we could include it as a fallback, and instruct folks in our docs to use their locally-installed `just` or `npx just` if they do not have it. 

I think this is perhaps not beneficial, but I still wanted to share the idea, or provide a place for contributors who do not have just to confirm that this is an issue worth solving.

**Comment by obulat** [comment]
>Perhaps, then, we could include it as a fallback, and instruct folks in our docs to use their locally-installed just or npx just if they do not have it.

Adding it to the docs as a possible work-around for not installing `just` sounds as a good solution. However, we should probably regularly be checking if the `npx just` works the same as the regular `just` locally.

**Reply by sarayourfriend** [reply]
Ideally we have 1 way we support of doing things, rather than multiple, especially not multiple that we wouldn't be regularly automatically testing. `just` is part of our CI script, so we always know it works, because if it breaks, literally everything else breaks :grinning:

**Comment by dhruvkb** [comment]
We have `just` listed as a hard requirement in our [general setup guide](https://docs.openverse.org/general/general_setup.html#just) and we link to 

- `just`'s installation section (which includes the [subsection for Node](https://github.com/casey/just#nodejs-installation))
- all the package managers that have `just` in their package list
- pre-built binaries for major OSes

I don't think we need to do anything else. Supporting multiple different ways to set up and invoke `just` will make it more difficult for us as almost all our docs have lots of `just` recipes.

**Reply by sarayourfriend** [reply]
I strongly agree with Dhruv. When `just` isn't available in a distro's package manager, it's available via cargo, which is easily installed and maintained using `rustup` in essentially all possible environments.

To me, the only reasonable alternative to our current usage of `just`, is to replace it with e.g., `invoke`, or some other tool within the rest of our toolchain. Python is a required dependency for Openverse development (our just recipes depend on it) so subsituting "Install `just`" with "Install `invoke` via `pipx`" would be a pretty straightforward change from a contributor perspective, without creating further oddities like the self-bootstrapping `pre-commit` we use currently (which I _always_ get mixed up with my global pre-commit install, which occasionally causes bizarre issues with `pre-commit`'s environment validation).

**Reply by AetherUnbound** [reply]
I agree with Dhruv as well. I've even gone and used the `musl` binaries myself, which worked on every POSIX system I interacted with - it should be straightforward to install and get running for folks.

**Comment by zackkrida** [comment]
Thanks for the suggestions here! It definitely makes sense to me to keep things as-is. I'm going to close this as I don't think any further discussion is necessary.
